extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -34,32 +37,69 @@ class AmpServiceWorkerInstallation extends AMP.BaseElement {
       return;
     }
     const src = this.element.getAttribute('src');
+    user.assert(src, 'src attribute is required in element %s', this.element);
     assertHttpsUrl(src, this.element);
 
+    /** @private {?string}  */
+    this.iframeSrc_ = null;
+
     if (isProxyOrigin(src) || isProxyOrigin(win.location.href)) {
+      const iframeSrc = this.element.getAttribute('data-iframe-src');
+      if (iframeSrc) {
+        assertHttpsUrl(iframeSrc, this.element);
+        const origin = parseUrl(iframeSrc).origin;
+        const docInfo = documentInfoFor(win);
+        const sourceUrl = parseUrl(docInfo.sourceUrl);
+        const canonicalUrl = parseUrl(docInfo.canonicalUrl);
+        user.assert(
+            origin == sourceUrl.origin ||
+            origin == canonicalUrl.origin,
+            'data-iframe-src (%s) should be a URL on the same origin as the ' +
+            'source (%s) or canonical URL (%s) of the AMP document.',
+            origin, sourceUrl.origin, canonicalUrl.origin);
+        this.iframeSrc_ = iframeSrc;
+        this.scheduleIframeLoad_();
+      }
       return;
     }
 
-    if (originMatches(win.location.href, src)) {
+    if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
       install(this.getWin(), src);
     } else {
-      if (getMode().development) {
-        user.warn(TAG,
-            'Did not install ServiceWorker because it does not ' +
-            'match the current origin: ' + src);
-      }
+      user.error(TAG,
+          'Did not install ServiceWorker because it does not ' +
+          'match the current origin: ' + src);
     }
   }
-}
 
-/**
- * Returns true if the 2 hrefs are on the same origin.
- * @param {string} href1
- * @param {string} href2
- * return {boolean}
- */
-function originMatches(href1, href2) {
-  return parseUrl(href1).origin == parseUrl(href2).origin;
+  /** @private */
+  scheduleIframeLoad_() {
+    viewerFor(this.getWin()).whenFirstVisible().then(() => {
+      // If the user is longer than 20 seconds on this page, load
+      // the external iframe to install the ServiceWorker. The wait is
+      // introduced to avoid installing SWs for content that the user
+      // only engaged with superficially.
+      return timer.promise(20000);
+    }).then(() => {
+      this.deferMutate(this.insertIframe_.bind(this));
```

Comment:
I this case I think we should also check for `viewer.isVisible()` before running mutate. Otherwise an article from less-intersting-publisher.com could be swiped away and sitting invisible while the reader is reading the super-interesting.publisher.com for the next 5 minutes.